python/lsst/daf/relation/_relation.py

```diff
@@ -0,0 +1,219 @@
+# This file is part of daf_relation.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+__all__ = ("Relation",)
+
+from abc import abstractmethod
+from typing import TYPE_CHECKING, AbstractSet, Generic, Iterable, TypeVar
+import json
+
+from ._columns import _T, UniqueKey
+from ._exceptions import ColumnError
+
+if TYPE_CHECKING:
+    from ._engines import EngineTag, EngineTree
+    from ._join_condition import JoinCondition
+    from ._order_by_term import OrderByTerm
+    from ._predicate import Predicate
+    from ._relation_visitor import _U, RelationVisitor
+
+
+_S = TypeVar("_S", bound="Relation")
+
+
+class Relation(Generic[_T]):
+    """An abstract interface for conceptual sets of tuples.
+
+    Relation is generic over the type used to represent its columns; see
+    `ColumnTag` for more information.
+
+    Notes
+    -----
+    The `Relation` class represents the core concept of `relational algebra`_:
+    a conceptual table with a well-defined set of columns and unique rows.  It
+    is rare for a Relation instance to actually correspond to a concrete
+    in-memory table, however; most derived Relation types actually represent an
+    operation on some other "base" relation or relations, forming an expression
+    tree that can be traversed by visitor classes (see `RelationVisitor`.
+
+    `Relation` is an unusual abstract base class in that the set of derived
+    types is closed to the `Leaf` class and the types in the
+    `~lsst.daf.relation.operations` subpackage; while external derived classes
```

Comment:
This attempt to link to the `operations` subpackage isn't rendering as a link via `package-docs build`.